File: leetcode/binary_search/774_Minimize_Max_Distance_to_Gas_Station/codes/bs.py
from typing import List

class Solution:
    def minmaxGasDist(self, stations: List[int], k: int) -> float:
        def is_possible(D):
            needed = 0
            for i in range(len(stations) - 1):
                dist = stations[i+1] - stations[i]
                needed += int(dist / D)
            return needed <= k

        lo, hi = 0, stations[-1] - stations[0]
        eps = 1e-6
        while hi - lo > eps:
            mid = (lo + hi) / 2
            if is_possible(mid):
                hi = mid
            else:
                lo = mid
        return lo

File: leetcode/binary_search/774_Minimize_Max_Distance_to_Gas_Station/codes/bs.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>

bool isPossible(int* stations, int size, int k, double D) {
    int count = 0;
    for (int i = 0; i < size - 1; ++i) {
        double dist = stations[i + 1] - stations[i];
        count += (int)(dist / D);
    }
    return count <= k;
}

double minmaxGasDist(int* stations, int stationsSize, int k) {
    double lo = 0, hi = stations[stationsSize - 1] - stations[0];
    double eps = 1e-6;

    while (hi - lo > eps) {
        double mid = (lo + hi) / 2;
        if (isPossible(stations, stationsSize, k, mid)) {
            hi = mid;
        } else {
            lo = mid;
        }
    }
    return lo;
}
